API review: XAML resource reference tracing (#8183)
microsoft/microsoft-ui-xaml: specs/xaml-resource-references-tracing-spec.md

Tracing XAML resource reference lookup failures
===

# Background

Failure to resolve a XAML resource reference (`{StaticResource}` or `{ThemeResource}`) is one of 

> the most common causes of app crashes. Such failures manifest as a stowed exception with the 

Reviewer: Stowed exception is mostly an internal thing, but I think this is true?

```suggestion
the most common causes of app crashes. Such failures manifest in the _native code_ debug output with the 
```

2. The referenced resource *does* exist in the app, but it is not reachable from the reference 
at run-time.

Unfortunately, the stowed exception message is the *only* information provided about the error, 
which makes it difficult or, more often, outright impossible to debug. This spec describes a new 
API on the `DebugSettings` class that developers can use to access more detailed information about 
the resource reference lookup failure.

# API Pages

_(Each of the following L2 sections correspond to a page that will be on docs.microsoft.com)_

## DebugSettings.XamlResourceReferenceFailed Event

Occurs when a [XAML resource reference](https://learn.microsoft.com/en-us/windows/apps/design/style/xaml-resource-dictionary) cannot be resolved.

_Spec note: is there a version of the 'ResourceDictionary and XAML resource references' article that 
links to the Windows App SDK documentation for relevant APIs?_

```c#
public event TypedEventHandler<DebugSettings,XamlResourceReferenceFailedEventArgs> 
XamlResourceReferenceFailed

```

### Remarks

`IsXamlResourceReferenceTracingEnabled` must be `true` and there must be a debugger attached to 

> the app process in order for `XamlResourceReferenceFailed` to fire and for tracing to appear in 

Reviewer: ```suggestion
the app process in order for `XamlResourceReferenceFailed` to be raised and for tracing to appear in 
```

> the app process in order for `XamlResourceReferenceFailed` to fire and for tracing to appear in 

Reviewer: If there's no debugger attached, the event isn't even raised? Why not? (Does BindingFailed behave that way?)

Author: BindingFailed behaves the same way. But you've got me thinking... designing it this way prevents a developer from logging the event information to their own reporting infrastructure. My original reasoning was that this is only useful as a debugging aid ergo without a debugger attached it's not useful but it's clear now that's faulty reasoning. I think it's worth making this event fire in all circumstances; BindingFailed is potentially noisy because it's non-fatal, but XamlResourceReferenceFailed doesn't have the same concern because the developer needs to go to some effort to make it non-fatal.

debugger output. You don't need to handle the event in order to see tracing appear in a debugger. 
The debugger output contains message information that goes to the **Output** window of the 
debugger. Attaching a `XamlResourceReferenceFailed` handler yourself is an advanced scenario for 

> when you want to see the raw message.

Reviewer: This section talks about debug output but doesn't describe that it must be the native debugger. It would help readability to move that description from the args to here.

## DebugSettings.IsXamlResourceReferenceTracingEnabled Property

> Gets or sets a value that indicates whether to engage the XAML resource reference tracing feature 
> of Microsoft Visual Studio when the app runs.

Reviewer: This isn't a VS feature is it? Windbg will show it too?

```suggestion
Gets or sets a value that indicates whether to raise the `XamlResourceReferenceFailed` event when a Xaml resource lookup fails.
```

    Finished searching merged dictionary with index '0'.
    Searching theme dictionary (active theme: 'Light') for resource with key 'OutputTextBlockStyl'.
      Searching dictionary 'ms-appx:///App.xaml' for resource with key 'OutputTextBlockStyl'.
      Finished searching dictionary 'ms-appx:///App.xaml'.
    Finished searching theme dictionary (active theme: 'Light').
  Finished searching dictionary 'ms-appx:///App.xaml'.
Finished search for resource with key 'OutputTextBlockStyl'.
```

## XamlResourceReferenceFailedEventArgs.Message Property

Gets the explanation of the XAML resource reference failure.

```c#
public string Message { get; }
```

# API Details
```c#
namespace Microsoft.UI.Xaml
{
  runtimeclass DebugSettings
  {
    // existing ...

    Boolean IsXamlResourceReferenceTracingEnabled;
    event Windows.Foundation.TypedEventHandler<Microsoft.UI.Xaml.DebugSettings,Microsoft.UI.Xaml.XamlResourceReferenceFailedEventArgs> XamlResourceReferenceFailed;
  };

  runtimeclass XamlResourceReferenceFailedEventArgs
  {
    XamlResourceReferenceFailedEventArgs();
    String Message{ get; };
  };
}

```

# Appendix

> This API is modeled on the existing `DebugSettings.BindingFailed` event. Like XAML resource references, classic Bindings cannot be modeled solely through static evaluation, e.g. at compile-time, and so a run-time event when a failure occurs is the best approach for providing developers with a means of determining the root cause.

Reviewer: Could you add this to the background? This is helpful for folks outside the team too in understanding the API design

APIs.

# API Pages

_(Updates to the 
[DebugSettings](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Xaml.DebugSettings)
class)_

## DebugSettings.XamlResourceReferenceFailed Event

Occurs when a
[XAML resource reference](https://learn.microsoft.com/en-us/windows/apps/design/style/xaml-resource-dictionary)
cannot be resolved.

_Spec note: This API is similar to the `DebugSettings.BindingFailed` event. Like XAML resource references, 
Bindings cannot be modeled solely through static evaluation, e.g. at compile-time, and so a run-time 
event when a failure occurs is the best approach for providing developers with a means of determining the 
root cause._

_Spec note: is there a version of the 'ResourceDictionary and XAML resource references' article that 
links to the Windows App SDK documentation for relevant APIs?_

```c#
public event TypedEventHandler<DebugSettings,XamlResourceReferenceFailedEventArgs> 
XamlResourceReferenceFailed

```

### Remarks

`IsXamlResourceReferenceTracingEnabled` must be `true` in order for this event to be raised.

> Error information is also logged to the native debug output,

Reviewer 2: Later text suggests that I have to set IsXamlResourceReferenceTracingEnabled = true to get native debug output.

> Error information is also logged to the native debug output if `IsXamlResourceReferenceTracingEnabled` is `true`,

so attaching a `XamlResourceReferenceFailed` handler yourself is an advanced scenario for 
getting the raw message programmatically.

## DebugSettings.IsXamlResourceReferenceTracingEnabled Property

Gets or sets a value indicating that when a XAML resource reference error occurs,
the `XamlResourceReferenceFailed` event should be raised,
and error information should be logged in the native debug output.

```c# 
public bool IsXamlResourceReferenceTracingEnabled { get; set; }
```

### Remarks

This property is `true` by default. When XAML resource reference tracing is enabled and you 
run your app with the native debugger attached, any  XAML resource reference errors appear 
in the **Output** window in Microsoft Visual Studio.

## XamlResourceReferenceFailedEventArgs Class

Provides event data for the `DebugSettings.XamlResourceReferenceFailed` event.

```c#
public sealed class XamlResourceReferenceFailedEventArgs
```

### Remarks

> `XamlResourceReferenceFailedEventArgs` is used for debugging XAML resource references. Wire the event 

Reviewer 2: ```suggestion
`XamlResourceReferenceFailedEventArgs` is used for debugging XAML resource references. Register the event 
```

> handler using `DebugSettings`, and use this data class as the result in your handler. You'll mainly be 

Reviewer 2: This data type is not the result. It is an input to the handler.
```suggestion
handler using `DebugSettings`. It will receive a reference to this class. You'll mainly be 
```

          Finished searching dictionary 'ms-appx:///Microsoft.UI.Xaml/Themes/themeresources.xaml'.
        Finished searching theme dictionary (active theme: 'Light').
      Finished searching dictionary 'ms-appx:///Microsoft.UI.Xaml/Themes/themeresources.xaml'.
    Finished searching merged dictionary with index '1'.
    Searching merged dictionary with index '0' for resource with key 'OutputTextBlockStyl'.
      Searching dictionary 'ms-appx:///ItemTemplates.xaml' for resource with key 'OutputTextBlockStyl'.
      Finished searching dictionary 'ms-appx:///ItemTemplates.xaml'.
    Finished searching merged dictionary with index '0'.
    Searching theme dictionary (active theme: 'Light') for resource with key 'OutputTextBlockStyl'.
      Searching dictionary 'ms-appx:///App.xaml' for resource with key 'OutputTextBlockStyl'.
      Finished searching dictionary 'ms-appx:///App.xaml'.
    Finished searching theme dictionary (active theme: 'Light').
  Finished searching dictionary 'ms-appx:///App.xaml'.
Finished search for resource with key 'OutputTextBlockStyl'.
```

> 

Reviewer 2: Presumably

> The above output is for illustrative purposes only. The precise format of the message is implementation-defined and may change in the future. Applications should not attempt to parse the message.

Author: Thanks for the suggestion! I like how you've phrased it.

> 

Reviewer 2: Is this string also logged as the RoOriginateError message string? If not, then it just flies into the ether and does not get captured in crash dumps, which makes it much less useful.

Author: The string gets saved in:
1. The "extraInfo" field of the XAML error context corresponding to the stowed exception
2. A per-thread object (this is primarily a fallback for #1)
3. XAML ETW trace provider

## XamlResourceReferenceFailedEventArgs.Message Property

> Gets the explanation of the XAML resource reference failure.

Reviewer 2: ```suggestion
A human-readable explanation of the XAML resource reference failure.
```

```c#
public string Message { get; }
```

# API Details
```c#
namespace Microsoft.UI.Xaml
{
  runtimeclass DebugSettings
  {
    // existing ...

    Boolean IsXamlResourceReferenceTracingEnabled;
    event Windows.Foundation.TypedEventHandler<Microsoft.UI.Xaml.DebugSettings,Microsoft.UI.Xaml.XamlResourceReferenceFailedEventArgs> XamlResourceReferenceFailed;
  };

  runtimeclass XamlResourceReferenceFailedEventArgs
  {

>     XamlResourceReferenceFailedEventArgs();

Reviewer: Is the public constructor necessary?

Author: No, this isn't expected to be constructed by the user.

and
[DebugSettings.IsBindingTracingEnabled](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Xaml.DebugSettings.IsBindingTracingEnabled)
APIs.

# API Pages

_(Updates to the 
[DebugSettings](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Xaml.DebugSettings)
class)_

## DebugSettings.XamlResourceReferenceFailed Event

Occurs when a
[XAML resource reference](https://learn.microsoft.com/en-us/windows/apps/design/style/xaml-resource-dictionary)
cannot be resolved.

_Spec note: This API is similar to the `DebugSettings.BindingFailed` event. Like XAML resource references, 
Bindings cannot be modeled solely through static evaluation, e.g. at compile-time, and so a run-time 
event when a failure occurs is the best approach for providing developers with a means of determining the 
root cause._

_Spec note: is there a version of the 'ResourceDictionary and XAML resource references' article that 
links to the Windows App SDK documentation for relevant APIs?_

```c#
public event TypedEventHandler<DebugSettings,XamlResourceReferenceFailedEventArgs> 
XamlResourceReferenceFailed

```

### Remarks

> `IsXamlResourceReferenceTracingEnabled` must be `true` in order for this event to be raised.

Reviewer 3: Is this different from BindingFailed?  This reads like the event will fire whether a debugger is present or not, unlike BindingFailed.

Reviewer 3: THe bindingFailed text reads: [IsBindingTracingEnabled](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.debugsettings.isbindingtracingenabled?view=windows-app-sdk-1.2#microsoft-ui-xaml-debugsettings-isbindingtracingenabled) must be true and there must be a debugger attached to the app process in order for BindingFailed to fire and for tracing to appear in debugger output.

Author: Yes, this was a change suggested by Mike and which I agree with; these errors are almost always fatal, unlike with Bindings, so (a) there's no risk of the event being noisy, and (b) requiring a debugger to be attached prevents the developer from sending the trace information to their own logging infrastructure.

Error information is also logged to the native debug output when the event is raised,
so attaching a `XamlResourceReferenceFailed` handler yourself is an advanced scenario for 
getting the raw message programmatically.

## DebugSettings.IsXamlResourceReferenceTracingEnabled Property

Gets or sets a value indicating that when a XAML resource reference error occurs,
the `XamlResourceReferenceFailed` event should be raised,
and error information should be logged in the native debug output.

```c# 
public bool IsXamlResourceReferenceTracingEnabled { get; set; }
```

### Remarks

This property is `true` by default. When XAML resource reference tracing is enabled and you 
run your app with the native debugger attached, any  XAML resource reference errors appear 
in the **Output** window in Microsoft Visual Studio.

## XamlResourceReferenceFailedEventArgs Class

Provides event data for the `DebugSettings.XamlResourceReferenceFailed` event.

```c#

> public sealed class XamlResourceReferenceFailedEventArgs

Reviewer 3: Does this also inherit from XamlEventArgs? NOTE: BindingFailedEventArgs docs say it inherits from object (which is incorrect).

Reviewer: Looks like a general doc issue with projections. It does derive from Object in the C# projection.

          Finished searching dictionary 'ms-appx:///Microsoft.UI.Xaml/Themes/themeresources.xaml'.
        Finished searching theme dictionary (active theme: 'Light').
      Finished searching dictionary 'ms-appx:///Microsoft.UI.Xaml/Themes/themeresources.xaml'.
    Finished searching merged dictionary with index '1'.
    Searching merged dictionary with index '0' for resource with key 'OutputTextBlockStyl'.
      Searching dictionary 'ms-appx:///ItemTemplates.xaml' for resource with key 'OutputTextBlockStyl'.
      Finished searching dictionary 'ms-appx:///ItemTemplates.xaml'.
    Finished searching merged dictionary with index '0'.
    Searching theme dictionary (active theme: 'Light') for resource with key 'OutputTextBlockStyl'.
      Searching dictionary 'ms-appx:///App.xaml' for resource with key 'OutputTextBlockStyl'.
      Finished searching dictionary 'ms-appx:///App.xaml'.
    Finished searching theme dictionary (active theme: 'Light').
  Finished searching dictionary 'ms-appx:///App.xaml'.
Finished search for resource with key 'OutputTextBlockStyl'.
```

## XamlResourceReferenceFailedEventArgs.Message Property

> A human-readable explanation of the XAML resource reference failure.

Reviewer 2: Maybe "A human-readable explanation (in English) of the..."

> A human-readable explanation of the XAML resource reference failure.

Author: Is it the norm for our docs to indicate when a developer-facing string is not localized? In other words, is the fact that *this* message isn't localized notable enough to call out (and if it is, should we be localizing it after all)?

> A human-readable explanation of the XAML resource reference failure.

Reviewer 2: I thought we localized even developer-facing strings. Because we have lots of developers who don't speak English natively.

> A human-readable explanation of the XAML resource reference failure.

Reviewer: No, we don't localize debug strings. That's actually part of why the error message in RoOriginate doesn't make it out into the Exception.Message of the C# projection; so that unlocalized strings don't accidentally sneak out.

There's little/no precedent in the docs for this because we don't usually doc the debug strings. I think Raymond's tweak is a good idea.

> A human-readable explanation of the XAML resource reference failure.

Author: Thanks for the explanation! I've updated the line to match Raymond's suggestion.

```c#
public string Message { get; }
```

# API Details
```c#
namespace Microsoft.UI.Xaml
{
  runtimeclass DebugSettings
  {
    // existing ...

    Boolean IsXamlResourceReferenceTracingEnabled;
    event Windows.Foundation.TypedEventHandler<Microsoft.UI.Xaml.DebugSettings,Microsoft.UI.Xaml.XamlResourceReferenceFailedEventArgs> XamlResourceReferenceFailed;
  };

>   runtimeclass XamlResourceReferenceFailedEventArgs

Reviewer 3: Should the class be sealed?

Reviewer: It is; in IDL runtime classes are sealed by default
